### ISSUE TITLE
Possible discrepancy between AdamW implementation and paper

### ISSUE THREAD

**Issue by ToucheSir** [deepmind/optax#292]
Per Algorithm 2:
![image](https://user-images.githubusercontent.com/4062143/151721233-548d3e25-5891-4571-a610-7f44f509b3a8.png)
The learning rate scaling is applied during the Adam step as opposed to after the weight decay as is [done now](https://github.com/deepmind/optax/blob/v0.1.0/optax/_src/alias.py#L270-L273). This is both unintuitive and not very clear since the paper opts to use `α` instead of `η` for the learning rate (`η` is reserved for an unrelated scheduled decay parameter), but it does appear to be what the authors intended. I believe swapping the order of `transform.add_decayed_weights(weight_decay, mask)` and `_scale_by_learning_rate(learning_rate)` in the chain would do the trick?

**Comment by n2cholas** [comment]
Wow, thanks for catching that nuance. A relevant excerpt of the paper here about the decoupling:
![image](https://user-images.githubusercontent.com/12474257/151860053-df82eb94-5a18-48e8-b24f-5cde3c441209.png)

Although swapping the order as you propose would give the user independent control of both the decay schedule and learning rate schedule, I'm not sure if it's the intuitive thing to do from an API perspective, as I believe most people expect the learning rate scaling to be the final step. Even after the swap, the API won't match the math given in the paper (as the paper uses three parameters as opposed to our two). As another data point, the [`PyTorch implementation`](https://pytorch.org/docs/stable/generated/torch.optim.AdamW.html#torch.optim.AdamW) makes the same design choice we do (learning rate applied last).

**Comment by ToucheSir** [comment]
Yes, the decay schedule would be yet another parameter and scaling, which I imagine most users don't even think about.

RE PyTorch, there is an issue at https://github.com/pytorch/pytorch/issues/38853. Reading the [linked discussion](https://github.com/pytorch/pytorch/pull/21250#issuecomment-520559993), this note stands out:

> I agree that this can be confusing, and the equivalent meaning of group['lr'] should be group['weight_decay'] = λη_t and not λ/α. This also means that when we introduce a weight scheduler in #22343, we can naturally schedule group['weight_decay'] independently of group['lr'].

My understanding of this is that the true `λ` may be larger than users expect with this implementation. How much that matters is another question, but it could be problematic for tuning hparams of specific optimizers. Unfortunately I could not find the justification behind this decision on the PyTorch side of things.

>  I believe most people expect the learning rate scaling to be the final step

Just to offer a counter-opinion, one argument of the AdamW paper is that the intuitive method of weight decay as an L2 regularization term isn't doing what people expect. Again, I'm not sure how much of an impact that has on real world performance 😄.

**Comment by mtthss** [comment]
Thanks everyone for their contribution to the discussion, overall we think it's best to stick with the current implementation which is consistent with what pytorch does (and what flax.optim used to do)
